### DISCUSSION TITLE
Delay when example.com is in ignore hosts

### DISCUSSION THREAD

**Discussion by rosydawn6** [mitmproxy/mitmproxy#5638] [Q&A] [answered]
I am seeing some timing delays depending on whether example.com is on or off the ignore hosts list.
When I startup mitmproxy 9.0.0 (detailed version info below)
mitmdump -v --listen-port 6077 --ignore-hosts 'example.com' --set console_eventlog_verbosity=info --set termlog_verbosity=info --set proxy_debug=true
and then issue
curl --proxy https://10.0.0.147:6077  https://example.com 

the curl command takes around 3 minutes to return.  This happens on both 9.0.0 and 8.1.1.  When I remove example.com from the ignore hosts list and start mitmproxy 9.0.0 like this:
mitmdump -v --listen-port 6077 --ignore-hosts 'examplea.com' --set console_eventlog_verbosity=info --set termlog_verbosity=info --set proxy_debug=true

followed by
curl --proxy https://10.0.0.147:6077  https://example.com 
the curl returns immediately.  Other domains that I have tested do not have this behavior.   Also if I start mimtproxy without any ignore hosts entry the curl also returns immediately.

Is this something peculiar to how example.com works or to my machine?  I have attached the output of the above transactions for version 9.0.0 when example.com is on/off the ignore hosts list if that may be of any use. outfile.9.0.0-onignore is the output when example.com is on the ignore list.  outfile.9.0.0-offignore is the output when examplea.com is on the ignore list (basically taking example.com off the list).  This behavior is also similar when running on 8.1.1.  I have not attached the 8.1.1 logs but can if it could help.

I am curious as to why it could be happening.  Thank you for any insights anyone may have!

Version 9.0.0 
Mitmproxy: 9.0.0.dev (+121, commit 5697b2e)
Python:    3.10.5
OpenSSL:   OpenSSL 3.0.5 5 Jul 2022
Platform:  Linux-5.13.0-52-generic-x86_64-with-glibc2.31

Mitmproxy: 8.1.1 binary
Python:    3.10.5
OpenSSL:   OpenSSL 3.0.3 3 May 2022
Platform:  Linux-5.13.0-52-generic-x86_64-with-glibc2.31

[outfile.9.0.0-onignore.txt](https://github.com/mitmproxy/mitmproxy/files/9768553/outfile.9.0.0-onignore.txt)
[outfile.9.0.0-offignore.txt](https://github.com/mitmproxy/mitmproxy/files/9768554/outfile.9.0.0-offignore.txt)

**Comment by mhils** [comment, marked as answer]
Thank you for the detailed logs. I'm not really sure what's going on here, both commands return quickly on my system. Looks like cURL is waiting for _something_ and then finally gives up when the connection is complete.

Some ideas on what you could try:

 - Does `curl -vvv [...]` turn up anything useful?
 - Does varying the HTTPS version in your cURL invocation change anything (`--http1.1`, `--http2`, etc)?
 - Do you see the same behavior when running mitmproxy as a reverse proxy? This would make it easy to peek into the TLS stream with SSLKEYLOGFILE.
 - Alternatively, maybe `--proxy http://10.0.0.147:6077/` (not `https://`) and then again WireShark + SSLKEYLOGFILE?

**Reply by rosydawn6** [reply]
Thanks for trying to replicate and for these suggestions.  They have helped clarify where the delay could be coming from - based on your second suggestion it seems to be related to the https version in curl.  Here are my results for your ideas.  One other thing I did not mention is that when I run the tests via a web browser I do not see the delay.   Anyway thanks for your help.  I vote to mark as closed.

1) I have attached the logs.  I do not see much here but maybe others may find some clues.  Perhaps it was evident from the prior logs I submitted, but I annotated in the curl verbose output file: what line the delay happens at -  it is the line before the last one.
noproxy.20221013T175649  (log without using mitmproxy)
noignore.http1.1.20221013T175832 (log when example.com is NOT on ignore list using http1.1 https version)
noignore.20221013T174516 (log when example.com is NOT on ignore list using default http2? https version
ignore.http1.1.20221013T174658 (log when example.com is ON ignore list using http1.1 https version
*** delay only happens here ignore.20221013T174713 (log when example.com is ON ignore list using default http2? https version) 
 
[noignore.http1.1.20221013T175832.txt](https://github.com/mitmproxy/mitmproxy/files/9787575/noignore.http1.1.20221013T175832.txt)
[noproxy.20221013T175649.txt](https://github.com/mitmproxy/mitmproxy/files/9787578/noproxy.20221013T175649.txt)
[ignore.20221013T174713.txt](https://github.com/mitmproxy/mitmproxy/files/9787579/ignore.20221013T174713.txt)
[ignore.http1.1.20221013T174658.txt](https://github.com/mitmproxy/mitmproxy/files/9787580/ignore.http1.1.20221013T174658.txt)
[noignore.20221013T174516.txt](https://github.com/mitmproxy/mitmproxy/files/9787581/noignore.20221013T174516.txt)

2) This was insightful
--http1.1 : no delay!!
--http2 : delay
--http2-prior-knowledge : delay
so it appears to be something with http2 https version.

3) I am not too familiar with setting this up.  Here is what I tried though I don't think it is correct.

mitmproxy side
mitmdump -v --listen-port 6077 --mode reverse:https://example.com --set console_eventlog_verbosity=info --set termlog_verbosity=info --set proxy_debug=true

On the curl side:
curl --proxy https://10.0.0.147:6077  https://example.com
curl: (60) SSL: no alternative certificate subject name matches target host name '10.0.0.147'
More details here: https://curl.haxx.se/docs/sslcerts.html

curl failed to verify the legitimacy of the server and therefore could not
establish a secure connection to it. To learn more about this situation and
how to fix it, please visit the web page mentioned above.

4) The curl returns without delay when using http even when example.com is in ignore host list so I did not yet go down this path.  I recall reading in the docs that for http transactions ignore hosts is not checked so maybe that is the reason.
